### ISSUE TITLE
Code completion stopped working for Haxe project after switching between tabs

### ISSUE THREAD

**Issue by piotrzarzycki21** [Moonshine-IDE/Moonshine-IDE#1212]
I'm struggling with issue in our haxe project [Super.Human.Installer](https://github.com/Moonshine-IDE/Super.Human.Installer). 
There is no consistent scenario which I could provide to describe issue, but in general:
1) Open project and one file - type in some places code with code completion help
2) Open another file and do exactly same
3) Open third file or get back to first one

Usually after couple of this operation code completion stop working and never get back until I restart project.

**Comment by joshtynjala** [comment]
No errors or anything in the console when this happens?

**Comment by rat-moonshine** [comment]
Code-completion stops working also as seen in VSCode - could it be something like that @joshtynjala (?) I remember you mentioned that there were some problem with Openfl/Lime side; I also wondering, if any fix is coming on that as well.

**Comment by joshtynjala** [comment]
When code completion stops working in VSCode, I think it's a Haxe compiler issue. I'm not aware of any coming fix for that. They might not even be aware of the issue, or if they are, how to reproduce it.

**Comment by piotrzarzycki21** [comment]
There is no error when it happens. However I'm not sure if language server stops actually. If I remember correctly (will check over weekend) - tooltips seems to be working still.

**Comment by joshtynjala** [comment]
If some code intelligence is still working, but some isn't, that seems more likely to be a Moonshine issue. Let me know what you find, and I'll try to take a look next week.

**Comment by piotrzarzycki21** [comment]
@joshtynjala I can confirm that language server is working when code completion stopped. I still doesn't have any further scenario wich can reproduce that issue. I have recorded short video where you see that I cannot launch code completion after some switching between tab and just writing code, using jump option. 

Link: https://1drv.ms/v/s!ApVpLyjpHDC2jrELsR61ZFs7-QGzFg?e=IsSQtM

**Comment by piotrzarzycki21** [comment]
This is my configuration:

```
Moonshine IDE™
Version 3.3.4, Build 3066, Player: 33.1(Harman)
Running on: Mac OS 13.4.1

Configured SDKs in Moonshine:
===========================================

Default SDK: Version 0.9 build 11-SNAPSHOT
Apache Flex SDK (Adobe AIR): Version 4.16.1 build 20171115, 32.0.0.116
Apache Flex SDK (Harman AIR): Version 4.16.1 build 20171115, 50.2.2.6
Apache Royale: Version 0.9 build 11-SNAPSHOT
Haxe: 4.2.5
Neko: 2.3.0
OpenJDK: openjdk version "17.0.5" 2022-10-18
OpenJDK 8: openjdk version "1.8.0_202"
Apache Ant: Apache Ant(TM) version 1.10.1 compiled on February 2 2017
Apache Maven: Apache Maven 3.6.0 (97c98ec64a1fdfee7767ce5ffb20918da4f719f3; 2018-10-24T20:41:47+02:00)
Gradle: Gradle 5.4.1
Grails: Not Installed
Feathers: Version 4.14.0 build 20190117, 32.0.0.116
Subversion: Not Installed
Git: git version 2.39.2 (Apple Git-143)
Node.js: Not Installed
HCL® Notes: Build 1101FP3_03312021 SHF15

MacPorts: Version: 2.8.1
Vagrant: Vagrant 2.3.7
VirtualBox: 7.0.8r156879
```

**Comment by joshtynjala** [comment]
> : Unable to build cache - completion features may be slower than expected. Try fixing the error (s) and restarting the language server.
> : Failed - try fixing the errors and restarting the language server:
> : Genesis/Source/genesis/application/GenesisApplication.hx:222: characters 9-24 : Tve not found : ApplicationMain

When you see this message, you should delete your project's output directory, and then close and reopen the project. It looks like you are using _Export_ as the name of the output directory. Sometimes it can be _bin_ instead.

Haxe's language server has some kind of build cache, and sometimes the data in the cache can be bad, and the language server doesn't know how to recover automatically.

It's likely that fixing the build cache issue will also resolve your issues with completion.

**Comment by joshtynjala** [comment]
> I have removed Export folder completely and nothing has change - I still have similar output and experience same issue.

Did you close and reopen the project after deleting the folder? (Or you can restart Moonshine instead)

> Why language server is reporting for example that non found ApplicationMain.

ApplicationMain.hx is a file that OpenFL generates as part of the build process. It's been my experience that this error is reported when there's a bad build cache.

**Comment by piotrzarzycki21** [comment]
> > I have removed Export folder completely and nothing has change - I still have similar output and experience same issue.
> 
> Did you close and reopen the project after deleting the folder? (Or you can restart Moonshine instead)
> 
> > Why language server is reporting for example that non found ApplicationMain.
> 
> ApplicationMain.hx is a file that OpenFL generates as part of the build process. It's been my experience that this error is reported when there's a bad build cache.

Yes I have restarted Moonshine. I'm not sure how to provide more technical details for this issue. The fact is that code completion stops working when I make couple of switches between files, type couple of line of codes. Sometimes it happened quick, sometimes it takes a little bit of time.

**Comment by joshtynjala** [comment]
> : Genesis/Source/genesis/application/GenesisApplication.hx:222: characters 9-24 : Type not found : ApplicationMain

This error is caused by `ApplicationMain.onException`. _ApplicationMain.hx_ is a template file, so the `ApplicationMain` class won't be available to code intelligence until you've successfully built the project at least once, and then restarted the language server. It doesn't get copied into your project until you build, so it's not a good idea to put public APIs like `onException` on `ApplicationMain` because it will cause compiler errors like this.

Today, I suggested to @rat-moonshine to move `onException` from `ApplicationMain` to another class that is on the classpath in Moonshine-IDE/HaxeTemplateApplication. The same change should happen in Moonshine-IDE/Super.Human.Installer. That should resolve the _Type not found : ApplicationMain_ error.

I suspect that it may help with the completion issue too. Since the language server is not starting properly, I don't trust it to work properly. However, please let me know if the completion issue persists after moving `onException` out of `ApplicationMain`.

**Comment by piotrzarzycki21** [comment]
> > : Genesis/Source/genesis/application/GenesisApplication.hx:222: characters 9-24 : Type not found : ApplicationMain
> 
> This error is caused by `ApplicationMain.onException`. _ApplicationMain.hx_ is a template file, so the `ApplicationMain` class won't be available to code intelligence until you've successfully built the project at least once, and then restarted the language server. It doesn't get copied into your project until you build, so it's not a good idea to put public APIs like `onException` on `ApplicationMain` because it will cause compiler errors like this.
> 
> Today, I suggested to @rat-moonshine to move `onException` from `ApplicationMain` to another class that is on the classpath in Moonshine-IDE/HaxeTemplateApplication. The same change should happen in Moonshine-IDE/Super.Human.Installer. That should resolve the _Type not found : ApplicationMain_ error.
> 
> I suspect that it may help with the completion issue too. Since the language server is not starting properly, I don't trust it to work properly. However, please let me know if the completion issue persists after moving `onException` out of `ApplicationMain` and you've restarted the project without any errors.

Thanks @joshtynjala this is definitely solution for this error with Type not found. However main issue is still there and it's happening even in a small project like Hello World, where I have couple of files - switching between them, typing couple of lines of code and  code completion stop working.

**Comment by joshtynjala** [comment]
> However main issue is still there and it's happening even in a small project like Hello World, where I have couple of files - switching between them, typing couple of lines of code and code completion stop working.

Okay, thanks. I'll investigate.

**Comment by joshtynjala** [comment]
It looks like the "unable to find document for URI" error was a separate issue, as I discovered it was triggered by closing a file (which you didn't do in the video).

I have found a separate issue in the `TextEditor` component where the completion list can be closed prematurely because a text line renderer isn't completely updated yet when trying to find the position of a character. That's probably the real issue that you are encountering. Working on a fix.
